File: source/python/topological_sort.py
from collections import deque

def topological_sort(adj, n):
    in_degree = [0] * n
    for u in range(n):
        for v in adj[u]:
            in_degree[v] += 1

    queue = deque()
    for i in range(n):
        if in_degree[i] == 0:
            queue.append(i)

    order = []
    while queue:
        u = queue.popleft()
        order.append(u)
        for v in adj[u]:
            in_degree[v] -= 1
            if in_degree[v] == 0:
                queue.append(v)

    return order

n = 6
adj = [[], [], [3], [1], [0, 1], [0, 2]]
result = topological_sort(adj, n)
print(f"topological order: {' '.join(str(x) for x in result)}")

File: source/rust/topological_sort.rs
use std::collections::VecDeque;

fn topological_sort(adj: &Vec<Vec<usize>>, n: usize) -> Vec<usize> {
    let mut in_degree = vec![0usize; n];
    for u in 0..n {
        for &v in &adj[u] {
            in_degree[v] += 1;
        }
    }
    let mut queue = VecDeque::new();
    for i in 0..n {
        if in_degree[i] == 0 {
            queue.push_back(i);
        }
    }
    let mut order = Vec::new();
    while let Some(u) = queue.pop_front() {
        order.push(u);
        for &v in &adj[u] {
            in_degree[v] -= 1;
            if in_degree[v] == 0 {
                queue.push_back(v);
            }
        }
    }
    order
}

fn main() {
    let adj = vec![vec![], vec![], vec![3], vec![1], vec![0, 1], vec![0, 2]];
    let result = topological_sort(&adj, 6);
    let s: Vec<String> = result.iter().map(|x| x.to_string()).collect();
    println!("topological order: {}", s.join(" "));
}
